### PR TITLE
Feature/prototype

### PR DESCRIPTION
**Pull request by stvnrs**
Initial prototyping completed - with a version in the PSGallery

### CODE REVIEW

**Comment by stvnrs** [comment]
@jamesfittes, @lfshr - package creation and testing work on this but publish is broken. Happy if you want to reject, I'll fix it and re-submit

**Comment by lfshr** [comment]
@stvnrs might as well update it and let us know when you think it's ready. The PR will keep track of changes you make on the branch.

**Comment by stvnrs** [comment]
I fixed it on the branch for issue 5 - I'll merge it into this  branch

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 [hunk 1]
```diff
@@ -0,0 +1,838 @@
...
+        $this.Message = $Message
+        $this.Severity = $Severity
+    }
+}
+
+function Get-AatWorkingFolder {
+    $ret = $Script:WorkingFolder 
+
+    if (-not $Script:WorkingFolder) {
+        $ret = $Pwd.Path
```
Comment (lfshr) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: Cheeky wee Get-Location workaround. 10/10, will not complain.

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 [hunk 2]
```diff
@@ -0,0 +1,838 @@
...
+        }
+        
+        Write-verbose -Message "Assets file: [$AssetsFileName]"
+        
+        $AssetsFolderPath = (Get-AatPackageFolderPath -Assets)
+        $AssetsFilePath = Join-Path -Path $AssetsFolderPath -ChildPath $AssetsFileName
+
+        Write-verbose -Message "Assets file path: '$AssetsFilePath'"
+
+        $Assets = Get-Content -Raw $AssetsFilePath | ConvertFrom-Json
```
Comment (lfshr) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: Positional parameter used

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 [hunk 3]
```diff
@@ -0,0 +1,838 @@
...
+    $RunbooksRoot = Join-Path -Path $Path -ChildPath 'runbooks';
+
+    Write-Verbose -Message "Runbooks Root: '$RunbooksRoot'"
+
+    if (-not (Test-Path $RunbooksRoot)) {
+        throw "Path not found: '$RunbooksRoot'"
+    }
+
+    Get-ChildItem -Path $RunbooksRoot -File -Filter $Filter | ForEach-Object {
+        $CommonParameters;
```
Comment (lfshr) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: Durty wee semi colons!

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 [hunk 4]
```diff
@@ -0,0 +1,838 @@
...
+    Write-Output "Filter: [$Filter]"
+
+    if (-not (Test-Path $AssetsRoot)) {
+        throw "Path not found: '$AssetsRoot'"
+    }
+    
+    Get-ChildItem -Path $AssetsRoot -File -Filter $Filter | ForEach-Object {
+        Write-Output "Assets File: '$($_.FullName)'"
+        #Write-Output (Get-Content $_.FullName -Raw) 
+        $Assets = Get-Content $_.FullName -Raw | ConvertFrom-Json;
```
Comment (lfshr) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: Positional parameter used

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 [hunk 5]
```diff
@@ -0,0 +1,838 @@
...
+
+    if (!Test-Path $ModulesRoot) {
+        throw "Path not found: '$ModulesRoot'"
+    }
+
+    Write-Output "Modules Root: '$ModulesRoot'"
+    
+    Get-ChildItem -Path $ModulesRoot -File -Filter '*.json' | ForEach-Object {
+        Write-Output "Modules file: '$($_.FullName)'";
+        $Modules = Get-Content $_.FullName -Raw | ConvertFrom-Json;
```
Comment (lfshr) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: Pos param

**Review by lfshr — APPROVED** [review]
Pretty awesome stuff! Approved with some suggestions.

Couple of discussion points:
* Since there are classes in play here, I think it's worth specifying 'PowerShellVersion = 5.0' in psd1 file.
* I think since this is public, we should support positional parameters. I know we hate them, but we should be accommodating.

File: CopyAndFormatMarkdownHelpfile.ps1 [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        $LineNumber++
+    }
+
+    if($LineNumber -eq $Lines.Count){
+        throw "Top level header not found in $SourceFilePath"
+    }
+}
+
+$Lines = $Lines | select -Skip $LineNumber #| % { if($_ -notmatch '^\s*$'){$_}}
```
Comment (stvnrs) [inline CopyAndFormatMarkdownHelpfile.ps1]: In a comment! :)

Comment (stvnrs) [inline CopyAndFormatMarkdownHelpfile.ps1]: Changed!

File: CopyAndFormatMarkdownHelpfile.ps1 [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        } 
+    
+        $PreviousLevel = $CurrentLevel
+    }    
+        
+    $LineNumber++
+}
+
+if($SkipEmptySecondLevelSections.IsPresent){
+    $NodesToRemove = $Nodes[0].ChildNodes| ? {-not $_.HasNonEmptyBodyText() -and -not $_.HasChildren()}
```
Comment (stvnrs) [inline CopyAndFormatMarkdownHelpfile.ps1]: and done,

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 (hunk 2 above)
Comment (stvnrs) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: on what line?

Comment (stvnrs) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: Oh. yeah - I see it! 

Done.

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 (hunk 3 above)
Comment (stvnrs) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: all gone!

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 (hunk 4 above)
Comment (stvnrs) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: done

File: src/PackageManagement/AzureAutomationTools.PackageManagement.psm1 (hunk 5 above)
Comment (stvnrs) [inline src/PackageManagement/AzureAutomationTools.PackageManagement.psm1]: and done

**Pull request merged**